scipy/stats/_discrepancy.pyx

```diff
@@ -0,0 +1,382 @@
+    .. [3] T. T. Warnock. "Computational investigations of low discrepancy
+       point sets". Applications of Number Theory to Numerical
+       Analysis, Academic Press, pp. 319-343, 1972.
+
+    Examples
+    --------
+    Calculate the quality of the sample using the discrepancy:
+
+    >>> from scipy.stats import qmc
+    >>> space = np.array([[1, 3], [2, 6], [3, 2], [4, 5], [5, 1], [6, 4]])
+    >>> l_bounds = [0.5, 0.5]
+    >>> u_bounds = [6.5, 6.5]
+    >>> space = qmc.scale(space, l_bounds, u_bounds, reverse=True)
+    >>> space
+    array([[0.08333333, 0.41666667],
+           [0.25      , 0.91666667],
+           [0.41666667, 0.25      ],
+           [0.58333333, 0.75      ],
+           [0.75      , 0.08333333],
+           [0.91666667, 0.58333333]])
+    >>> qmc.discrepancy_vect(space)
+    0.008142039609053464
+
+    We can also compute iteratively the ``CD`` discrepancy by using
+    ``iterative=True``.
+
+    >>> disc_init = qmc.discrepancy_vect(space[:-1], iterative=True)
+    >>> disc_init
+    0.04769081147119336
+    >>> qmc.update_discrepancy(space[-1], space[:-1], disc_init)
+    0.008142039609053513
+
+    """
+    sample = np.asarray(sample, dtype=np.float64)
+
+    # Checking that sample is within the hypercube and 2D
+    if not sample.ndim == 2:
+        raise ValueError('Sample is not a 2D array')
+
+    if not (np.all(sample >= 0) and np.all(sample <= 1)):
+        raise ValueError('Sample is not in unit hypercube')
+
+    if method == 'CD':
+        return centered_discrepancy(sample, iterative)
+
+    elif method == 'WD':
+        return wrap_around_discrepancy(sample, iterative)
+
+    elif method == 'MD':
+        return mixture_discrepancy(sample, iterative)
+
+    elif method == 'L2-star':
+        return l2_star_discrepancy(sample, iterative)
+    else:
+        raise ValueError('{} is not a valid method. Options are '
+                     'CD, WD, MD, L2-star.'.format(method))
+
+
+cdef double centered_discrepancy(double[:, ::1] sample_view,
+                             bint iterative):
+    cdef:
+        Py_ssize_t n = sample_view.shape[0]
+        Py_ssize_t d = sample_view.shape[1]
+        Py_ssize_t i = 0, j = 0, k = 0
+        double prod = 1, disc1 = 0
+
+    for i in range(n):
+        for j in range(d):
+                prod *= (
+                        1 + 0.5 * fabs(sample_view[i, j] - 0.5) - 0.5 * fabs(
+                        sample_view[i, j] - 0.5) ** 2
+                )
+        disc1 += prod
+        prod = 1
+
+    cdef:
+        double disc2 = 0
+
+    for i in range(n):
+        for j in range(n):
+            for k in range(d):
+                    prod *= (
+                            1 + 0.5 * fabs(sample_view[i, k] - 0.5)
+                            + 0.5 * fabs(sample_view[j, k] - 0.5)
+                            - 0.5 * fabs(sample_view[i, k] - sample_view[j, k])
+                    )
+            disc2 += prod
+            prod = 1
```

Comment:
As this loop is quite long, did you try with `prange` from `cython.parallel`? In this case, do not forget the `openmp` flag in compilation.

Reply:
note openmp is not used in scipy, see [this thread](https://mail.python.org/pipermail/scipy-dev/2019-August/023635.html). However testing it can be a usefull information if parallelism can improve the performances.

Reply:
Thank you @jb-leger. I have some issue with fopenm, but cython seems to compile nonetheless with `orange`. The result are quite unbelievable for the centered discrepancy (I only tested this one since openmp is not used in scipy anyway). Here are the result : 
||`old scipy time`|`cython time`|`scipy / cython`|
|---|---|---|---|
|time for `n=100` & `d=2`|0.0158s|0.0022s|7.1422|
|time for `n=100` & `d=10`|0.0589s|0.0018s|33.4766|
|time for `n=1000` & `d=10`|3.8109s|0.0065s|584.6393|

Reply:
Ok, this can be parallelized, with a `10000×100` sample (the same in the two test), without paralellism:

```
In [4]: %time stats.qmc.discrepancy(sample)
CPU times: user 18.6 s, sys: 0 ns, total: 18.6 s
Wall time: 18.6 s
Out[4]: 489346.18589523673
```
With parallelism:
```
In [13]: %time _discrepancy.discrepancy(sample)
CPU times: user 33.1 s, sys: 44.7 ms, total: 33.2 s
Wall time: 4.25 s
Out[13]: 489346.1858951062
```

Reply:
I will have a look how to do that directly with `pthread`. @V0lantis, I will send you results (and code) of my tests.

Reply:
Can I parallelize without openmp?

Reply:
Only for your information, when you use prange, have a look on the c code. You will see a `#pragma omp parallel`, you must check a reduction is recognized with the operator sum. And you must also check the initialization of private (thread dependant) variable, in your particular case the `prod` must be initialized before.

Reply:
isn't it already the case? (line 158 or line 149)